noir-projects/noir-contracts/contracts/fpc_contract/src/main.nr

```diff
@@ -16,61 +20,115 @@ contract FPC {
+    ///
+    /// ## Overview
+    /// The refund flow works as follows:
+    /// 1. We pull the `max_fee` from the user's balance of the AA to this contract (requires setting an authwit),
+    /// 2. then the private and public functions of a tx get executed,
+    /// 3. at this point we know the tx fee so we can compute how much of AA the user needs to pay to FPC and how much
+    /// of it will be refunded back. Note that this is computed based on an exchange rate between AA and fee juice.
+    /// 4. the protocol deducts the actual fee denominated in fee juice from the FPC's balance.
     #[private]
-    fn fee_entrypoint_public(amount: Field, asset: AztecAddress, nonce: Field) {
-        FPC::at(context.this_address())
-            .prepare_fee(context.msg_sender(), amount, asset, nonce)
+    fn fee_entrypoint_public(max_fee: Field, nonce: Field) {
+        // TODO(PR #8022): Once PublicImmutable performs only 1 merkle proof here, we'll save ~4k gates
+        let config = storage.config.read();
+
+        // We pull the max fee from the user's balance of the accepted asset to this contract.
+        // docs:start:public_call
+        Token::at(config.accepted_asset)
+            .transfer_in_public(context.msg_sender(), context.this_address(), max_fee, nonce)
             .enqueue(&mut context);
+        // docs:end:public_call
+
         context.set_as_fee_payer();
         // TODO(#6277) for improving interface:
-        // FPC::at(context.this_address()).pay_refund(context.msg_sender(), amount, asset).set_public_teardown_function(&mut context);
+        // FPC::at(context.this_address()).pay_refund(...).set_public_teardown_function(&mut context);
         context.set_public_teardown_function(
             context.this_address(),
             comptime { FunctionSelector::from_signature("pay_refund((Field),Field,(Field))") },
-            [context.msg_sender().to_field(), amount, asset.to_field()],
+            [context.msg_sender().to_field(), max_fee, config.accepted_asset.to_field()],
         );
     }
 
+    /// Pays the refund to the `refund_recipient`. The refund is the difference between the `max_fee` and
+    /// the actual fee. `accepted_asset` is the asset in which the refund is paid. It's passed as an argument
+    /// to avoid the need for another read from public storage.
     #[public]
     #[internal]
-    fn prepare_fee(from: AztecAddress, amount: Field, asset: AztecAddress, nonce: Field) {
-        // docs:start:public_call
-        Token::at(asset).transfer_in_public(from, context.this_address(), amount, nonce).call(
-            &mut context,
-        );
-        // docs:end:public_call
+    fn pay_refund(refund_recipient: AztecAddress, max_fee: Field, accepted_asset: AztecAddress) {
+        let actual_fee = context.transaction_fee();
+        assert(!max_fee.lt(actual_fee), "Max fee paid to the paymaster does not cover actual fee");
+        let refund = max_fee - actual_fee;
```

Comment:
Iiuc:
The `max_fee` that we lined up in the earlier function is denominated in AA.
The `actual_fee` that we're grabbing from context.transaction_fee() is denominated in fee juice.
So we shouldn't be doing arithmetic on two different units of currency. Shouldn't there be an "exchange" step, at some point, to convert `actual_fee` from fee juice to AA? 

Or am I misunderstanding what `context.transaction_fee()` does?

Reply:
You are absolutely right and currently the exchange rate is fixed which is super confusing. See [this comment](https://aztecprotocol.slack.com/archives/C04PUD9AA4W/p1733861678534879) on slack where I proposed to try to come up with a more real exchange rate mechanism.

Reply:
Created [an issue](https://github.com/AztecProtocol/aztec-packages/issues/10805) and linked it in the code